include/revng/Model/model-schema.yml

```diff
@@ -245,469 +349,915 @@ definitions:
           elementType: StructField
-  - name: Type
-    doc: |-
-      A type such as an array, a pointer, a primitive or a defined type.
+
+  - name: StructField
     type: struct
+
+    doc: |-
+      A field of a `StructDefinition`.
+
+      It is composed by the offset of the field and its type.
+
+    key:
+      - Offset
+
     fields:
-      - name: Kind
-        type: TypeKind
-      - name: IsConst
-        type: bool
+      - name: Offset
+        doc: The number of bytes in the given struct before this field.
+        type: uint64_t
+
+      - name: Name
+        doc: The name of this field.
+        type: string
         optional: true
-    abstract: true
-  - name: PrimitiveType
-    type: struct
-    inherits: Type
-    fields:
-      - name: PrimitiveKind
-        type: PrimitiveKind
-      - name: Size
-        doc: |
-          As of now, for floating point primitives, supported sizes include:
-          ```
-          { 2, 4, 8, 10, 12, 16 }
-          ```
-          For non-floating point, they are:
-          ```
-          { 1, 2, 4, 8, 16 }
-          ```
 
-          Note that `Void` *must* have size of 0 while `Generic` can use all the
-          supported sizes, floating point or not.
-        type: uint64_t
+      - name: Comment
+        doc: The comment that will be emitted before this field's *definition*.
+        type: string
         optional: true
-  - name: PointerType
-    type: struct
-    inherits: Type
-    doc: A pointer type.
-    fields:
-      - name: PointerSize
-        doc: As of now, only 4 and 8 byte pointers are supported.
-        type: uint64_t
-      - name: PointeeType
+
+      - name: Type
+        doc: The type of the field.
         type: Type
-        doc: The pointee type.
         upcastable: true
-  - name: DefinedType
-    type: struct
-    inherits: Type
-    doc: A reference to a `TypeDefinition`.
-    fields:
-      - name: Definition
-        reference:
-          pointeeType: TypeDefinition
-          rootType: Binary
-  - name: ArrayType
+
+  - name: UnionDefinition
+    inherits: TypeDefinition
     type: struct
+
     doc: |-
-      An array of `Type`s.
-    inherits: Type
+      A `union` type.
+
     fields:
-      - name: ElementCount
-        type: uint64_t
-        doc: The number of elements.
-      - name: ElementType
-        type: Type
-        upcastable: true
-        doc: The type of the elements of the array.
+      - name: Fields
+        doc: The list of alternative types in this `union`.
+        sequence:
+          type: SortedVector
+          elementType: UnionField
+
   - name: UnionField
+    type: struct
+
     doc: |-
-      An alternative of a `UnionDefinition`.
+      Specifies one of the possible types a given `UnionDefinition` can assume.
 
       It is composed by a index and the `Type`.
-    type: struct
+
+    key:
+      - Index
+
     fields:
       - name: Index
+        doc: A numeric index unique within a given `UnionDefinition`.
```

Comment:
Shall we add a not that these need to be *dense* (i.e., no holes)?

Reply:
I went with:
```
          A numeric index unique within a given `UnionDefinition`.

          Note that indexing is dense, so if a given union has a type with
          index 4, it must also have types indexed 0-3.
```